bindings/matrix-sdk-ffi/src/client_builder.rs

```diff
@@ -0,0 +1,87 @@
+use std::{fs, path::PathBuf, sync::Arc};
+
+use anyhow::Context;
+use matrix_sdk::{
+    ruma::UserId, store::make_store_config, Client as MatrixClient,
+    ClientBuilder as MatrixClientBuilder,
+};
+use sanitize_filename_reader_friendly::sanitize;
+
+use super::{client::Client, ClientState, RUNTIME};
+
+#[derive(Clone)]
+pub struct ClientBuilder {
+    base_path: Option<String>,
+    username: Option<String>,
+    homeserver_url: Option<String>,
+    inner: MatrixClientBuilder,
+}
+
+impl ClientBuilder {
+    pub fn new() -> Self {
+        Self {
+            base_path: None,
+            username: None,
+            homeserver_url: None,
+            inner: MatrixClient::builder().user_agent("rust-sdk-ios"),
+        }
+    }
+
+    pub fn base_path(self: Arc<Self>, path: String) -> Arc<Self> {
+        let mut builder = unwrap_or_clone_arc(self);
+        builder.base_path = Some(path);
+        Arc::new(builder)
+    }
+
+    pub fn username(self: Arc<Self>, username: String) -> Arc<Self> {
+        let mut builder = unwrap_or_clone_arc(self);
+        builder.username = Some(username);
```

Comment:
Why hold an inner builder but still have all parameters twice? Why not just 
```suggestion
        builder.inner = builder.inner.username(username);
```

For the homeserver, too...

Reply:
We can't currently read those fields from the inner builder, but they're needed in `build`. I think there's room for improvement here, but it's not entirely clear how this should be improved. In general I think the FFI layer shouldn't add its own logic like deriving a store path from a base path plus username / homeserver, but doing that immediately would mean extra work for using the FFI compared to before this change, which is also not great.